lite/dbprovider.go

```diff
@@ -0,0 +1,260 @@
+	dbp.limit = limit
+	return dbp
+}
+
+// Implements PersistentProvider.
+func (dbp *DBProvider) SaveFullCommit(fc FullCommit) error {
+
+	dbp.logger.Info("DBProvider.SaveFullCommit()...", "fc", fc)
+	batch := dbp.db.NewBatch()
+
+	// Save the fc.validators.
+	// We might be overwriting what we already have, but
+	// it makes the logic easier for now.
+	vsKey := validatorSetKey(fc.ChainID(), fc.Height())
+	vsBz, err := dbp.cdc.MarshalBinary(fc.Validators)
+	if err != nil {
+		return err
+	}
+	batch.Set(vsKey, vsBz)
+
+	// Save the fc.NextValidators.
+	nvsKey := validatorSetKey(fc.ChainID(), fc.Height()+1)
+	nvsBz, err := dbp.cdc.MarshalBinary(fc.NextValidators)
+	if err != nil {
+		return err
+	}
+	batch.Set(nvsKey, nvsBz)
+
+	// Save the fc.SignedHeader
+	shKey := signedHeaderKey(fc.ChainID(), fc.Height())
+	shBz, err := dbp.cdc.MarshalBinary(fc.SignedHeader)
+	if err != nil {
+		return err
+	}
+	batch.Set(shKey, shBz)
+
+	// And write sync.
+	batch.WriteSync()
+
+	// Garbage collect.
+	// TODO: optimize later.
+	if dbp.limit > 0 {
+		dbp.deleteAfterN(fc.ChainID(), dbp.limit)
+	}
+
+	return nil
+}
+
+// Implements Provider.
+func (dbp *DBProvider) LatestFullCommit(chainID string, minHeight, maxHeight int64) (
+	FullCommit, error) {
+
+	dbp.logger.Info("DBProvider.LatestFullCommit()...",
+		"chainID", chainID, "minHeight", minHeight, "maxHeight", maxHeight)
+
+	if minHeight <= 0 {
+		minHeight = 1
+	}
+	if maxHeight == 0 {
+		maxHeight = 1<<63 - 1
+	}
+
+	itr := dbp.db.ReverseIterator(
+		signedHeaderKey(chainID, maxHeight),
+		signedHeaderKey(chainID, minHeight-1),
+	)
+	defer itr.Close()
+
+	for itr.Valid() {
+		key := itr.Key()
+		_, _, ok := parseSignedHeaderKey(key)
+		if !ok {
+			// Skip over other keys.
+			itr.Next()
+			continue
+		} else {
+			// Found the latest full commit signed header.
+			shBz := itr.Value()
+			sh := types.SignedHeader{}
+			err := dbp.cdc.UnmarshalBinary(shBz, &sh)
+			if err != nil {
+				return FullCommit{}, err
+			} else {
+				lfc, err := dbp.fillFullCommit(sh)
+				if err == nil {
+					dbp.logger.Info("DBProvider.LatestFullCommit() found latest.", "height", lfc.Height())
+					return lfc, nil
+				} else {
+					dbp.logger.Info("DBProvider.LatestFullCommit() got error", "lfc", lfc)
+					dbp.logger.Info(fmt.Sprintf("%+v", err))
+					return lfc, err
+				}
+			}
+		}
+	}
+	return FullCommit{}, lerr.ErrCommitNotFound()
+}
+
+func (dbp *DBProvider) ValidatorSet(chainID string, height int64) (valset *types.ValidatorSet, err error) {
+	return dbp.getValidatorSet(chainID, height)
+}
+
+func (dbp *DBProvider) getValidatorSet(chainID string, height int64) (valset *types.ValidatorSet, err error) {
+	vsBz := dbp.db.Get(validatorSetKey(chainID, height))
+	if vsBz == nil {
+		err = lerr.ErrMissingValidators(chainID, height)
+		return
+	}
+	err = dbp.cdc.UnmarshalBinary(vsBz, &valset)
+	if err != nil {
+		return
+	}
+	valset.TotalVotingPower() // to test deep equality.
```

Comment:
Can you explain this line?

Reply:
I'll comment.  This makes it easier to test for e.g. valset equivalence using assert.Equal (tests for deep equality) in our tests, which also tests for unexported/private field equivalence.